hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java

```diff
@@ -231,77 +275,120 @@ private AddressBook parse(FileData fileData) {
-     * AddressBook
+     * Get Node id and account id
      *
-     * @param nodeAddressBook    node address book proto
-     * @param consensusTimestamp transaction consensusTimestamp
-     * @return
+     * @param nodeAddressProto
+     * @return Pair of nodeId and nodeAccountId
      */
-    private Collection<AddressBookEntry> retrieveNodeAddressesFromAddressBook(NodeAddressBook nodeAddressBook,
-                                                                              long consensusTimestamp) {
-        ImmutableList.Builder<AddressBookEntry> builder = ImmutableList.builder();
-        Set<Long> nodeIdSet = new HashSet<>();
+    private static Pair<Long, EntityId> getNodeIds(NodeAddress nodeAddressProto) {
+        var memo = nodeAddressProto.getMemo().toStringUtf8();
+        EntityId memoNodeAccountId = StringUtils.isEmpty(memo) ? EntityId.EMPTY : EntityId
+                .of(memo, EntityTypeEnum.ACCOUNT);
+        var nodeAccountId = nodeAddressProto.hasNodeAccountId() ? EntityId
+                .of(nodeAddressProto.getNodeAccountId()) : memoNodeAccountId;
+
+        var nodeId = nodeAddressProto.getNodeId();
+        // ensure valid nodeId. In early versions of initial addressBook (entityNum < 20) all nodeIds are set to 0
+        if (nodeId == 0 && nodeAccountId.getEntityNum() < 20 &&
+                nodeAccountId.getEntityNum() != INITIAL_NODE_ID_ACCOUNT_ID_OFFSET) {
+            nodeId = nodeAccountId.getEntityNum() - INITIAL_NODE_ID_ACCOUNT_ID_OFFSET;
+        }
 
-        for (NodeAddress nodeAddressProto : nodeAddressBook.getNodeAddressList()) {
-            int port = nodeAddressProto.getPortno();
-            AddressBookEntry addressBookEntry = AddressBookEntry.builder()
-                    .consensusTimestamp(consensusTimestamp)
-                    .memo(nodeAddressProto.getMemo().toStringUtf8())
-                    .ip(nodeAddressProto.getIpAddress().toStringUtf8())
-                    .port(port == 0 ? null : port)
-                    .publicKey(nodeAddressProto.getRSAPubKey())
-                    .nodeCertHash(nodeAddressProto.getNodeCertHash().toByteArray())
-                    .nodeId(nodeAddressProto.getNodeId())
-                    .nodeAccountId(EntityId.of(nodeAddressProto.getNodeAccountId()))
-                    .build();
-            builder.add(addressBookEntry);
-            nodeIdSet.add(nodeAddressProto.getNodeId());
+        return Pair.of(nodeId, nodeAccountId);
+    }
+
+    private static AddressBookEntry getAddressBookEntry(NodeAddress nodeAddressProto, long consensusTimestamp,
+                                                        Pair<Long, EntityId> nodeIds) {
+        return AddressBookEntry.builder()
+                .id(new AddressBookEntry.Id(consensusTimestamp, nodeIds.getLeft()))
+                .description(nodeAddressProto.getDescription())
+                .memo(nodeIds.getRight().entityIdToString())
```

Comment:
We should just directly copy the memo instead of assuming it is a node account ID. In the future, it will not be populated and we should just have an empty string.
```suggestion
                .memo(nodeAddressProto.getMemo().toStringUtf8())
```

Reply:
Also double check what happens if it is empty. Can we still call toStringUtf8()?

Reply:
True, at this stage we don't need to parse the memo so we can take as is.
When it's empty we just get an empty string. So we're all good

Reply:
Note that the field is `bytes` not a `string`. So it's not guaranteed to be empty string in the future. I would add some extra validation around null/empty.

Reply:
Actually proto default is `ByteString.EMPTY`. Did add a test to surface this for when it's not set.
However, will add a check for if it's explicitly set to null